[NVPTX] Constant fold NVVM fmin and fmax

LewisCrawford opened the pull request:
Add constant-folding for nvvm float/double fmin + fmax intrinsics, including all combinations of xorsign.abs, nan-propagation, and ftz.

LewisCrawford committed:
[NVPTX] Constant fold NVVM fmin and fmax

Add constant-folding for nvvm float/double fmin + fmax intrinsics,
including all combinations of xorsign.abs, nan-propagation, and ftz.

Artem-B commented on llvm/include/llvm/IR/NVVMIntrinsicUtils.h:
```diff
@@ -38,9 +38,8 @@ enum class TMAReductionOp : uint8_t {
 };
 
-inline bool IntrinsicShouldFTZ(Intrinsic::ID IntrinsicID) {
+inline bool FloatToIntIntrinsicShouldFTZ(Intrinsic::ID IntrinsicID) {
```
Nit. The new name implies that the function is intended for f2i intrinsics only, yet it will still happily accept other intrinsics.

If it is intended to be restricted to particular subset of intrinsics (as opposed to telling us whether a given intrinsic folds to zero), then it should handle all entries in the set, and have an assertion in case an unexpected intrinsic is passed.

I'd just revert to the old name, and, possibly, extend the list of intrinsics, if we're missing any other FTZ variants.

LewisCrawford replied:
This is mostly in response to @AlexMaclean 's comment on #118965 here:  https://github.com/llvm/llvm-project/pull/118965#discussion_r1884210682 , where he suggested it would be better to have helper functions for each class of intrinsic, rather than one big switch statement with 176 unrelated intrinsics to check for rounding modes, and 144 unrelated intrinsics with the FTZ modifier. Here, I'm adding a new FMinFMaxShouldFTZ function, and renaming IntrinsicShouldFTZ to FloatToIntIntrinsicShouldFTZ instead of combining both into a single switch statement.

I've changed these helper functions all to have explicit checks for all entries in the sets now, which means there are roughly double the number of case statements needed in this file.

Hopefully these rather unwieldy case statements can help motivate the use of flag arguments in future intrinsics, such as in #121507

Artem-B replied:
I do agree with Alex' suggestion, just pointing out that the name of the function should match what it does. Right now it does not.

> I've changed these helper functions all to have explicit checks for all entries in the sets now, which means there are roughly double the number of case statements needed in this file.

I would say that those would be there for a reason. they wold also make it possible to add an assertion that we're not using them unintentionally for an intrinsic they were not intended for, which is too easy to do with intrinsic identifiers.

Artem-B commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -1689,6 +1689,28 @@ bool llvm::canConstantFoldCallTo(const CallBase *Call, const Function *F) {
   case Intrinsic::x86_avx512_cvttsd2usi64:
     return !Call->isStrictFP();
 
+  // NVVM FMax intrinsics
+  case Intrinsic::nvvm_fmax_d:
+  case Intrinsic::nvvm_fmax_f:
+  case Intrinsic::nvvm_fmax_ftz_f:
+  case Intrinsic::nvvm_fmax_ftz_nan_f:
+  case Intrinsic::nvvm_fmax_ftz_nan_xorsign_abs_f:
+  case Intrinsic::nvvm_fmax_ftz_xorsign_abs_f:
+  case Intrinsic::nvvm_fmax_nan_f:
+  case Intrinsic::nvvm_fmax_nan_xorsign_abs_f:
+  case Intrinsic::nvvm_fmax_xorsign_abs_f:
+
+  // NVVM FMin intrinsics
+  case Intrinsic::nvvm_fmin_d:
+  case Intrinsic::nvvm_fmin_f:
+  case Intrinsic::nvvm_fmin_ftz_f:
+  case Intrinsic::nvvm_fmin_ftz_nan_f:
+  case Intrinsic::nvvm_fmin_ftz_nan_xorsign_abs_f:
+  case Intrinsic::nvvm_fmin_ftz_xorsign_abs_f:
+  case Intrinsic::nvvm_fmin_nan_f:
+  case Intrinsic::nvvm_fmin_nan_xorsign_abs_f:
+  case Intrinsic::nvvm_fmin_xorsign_abs_f:
```
Should we follow the folding behavior of the builtin min/max intrinsics based on `isStrictFP()` ? Granted, "FP environment" on the GPU is effectively fixed, but it may be prudent to make the folding decisions consistently across host and GPU compilations.

LewisCrawford replied:
I'd prefer to just always allow these to be folded, rather than rely on the `isStrictFP()` flag, because as you say, the FP environment should be effectively fixed. The semantics of `llvm.minimum` and `nvvm_fmin` are slightly different too, so they already cannot be used interchangably between host and device, so having slightly different constant-folding behavior for both would make sense.

Artem-B replied:
OK.

Artem-B commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -2955,6 +3014,79 @@ static Constant *ConstantFoldIntrinsicCall2(Intrinsic::ID IntrinsicID, Type *Ty,
+      case Intrinsic::nvvm_fmin_nan_xorsign_abs_f:
+      case Intrinsic::nvvm_fmin_xorsign_abs_f: {
+
+        bool ShouldCanonicalizeNaNs = IntrinsicID != Intrinsic::nvvm_fmax_d &&
+                                      IntrinsicID != Intrinsic::nvvm_fmin_d;
```
Nit: `not this and not that` is harder to read than `not (this or that)`. The former case operates on two possibly infinite sets of "not X", while the latter is a 2-element set and the intent is immediately obvious.

Artem-B reviewed (commented):
LGTM overall.

Artem-B commented:
On a side note, would it make sense to autoupgrade some of the min/max intrinsics to their LLVM counterpart, so LLVM can optimize them?

LewisCrawford committed:
Make intrinsic helpers accept closed sets

Make all the helper functions in NVVMIntrinsicUtils.h explicitly
accept all valid intrinsics, and call llvm_unreachable for any
unexpected intrinsics.

Rename some f2i/d2i helpers to make it clearer their scope is for finite
groups of intrinsics, rather than all intrinsics.

LewisCrawford committed:
Refactor if condition for readability

LewisCrawford commented:
I believe the issue with auto-upgrading these (even the ones without ftz/xorsignabs/nan modifiers) is that the semantics are slightly different than the existing `llvm.minnum` and `llvm.minimum intrinsics`.

`nvvm.fmin` maps to `minimumNumber` from IEEE754:2019 (it guarantees -0 < +0, and if either operand is a NaN it returns the other operand). It seems like support for this was recently added to LLVM's libc in https://github.com/llvm/llvm-project/pull/86016 , but there is not support for it with an LLVM IR intrinsic yet.

LewisCrawford committed:
Rename FloatToInt intrinsics to FPToInteger

Try to avoid the connotation that FloatToInt was restricted to f2i
intrinsics, in the hopes that FPToInteger covers a borader range of
floating-point types (float + double), and both signed and unsigned
integers with both 32 and 64 bits (rather than just specifically
C-style floats and ints).

Artem-B commented on llvm/test/Transforms/InstSimplify/const-fold-nvvm-fmin-fmax.ll:
```diff
@@ -0,0 +1,614 @@
+; CHECK-LABEL: define float @test_fmax_1_25_neg_2_ftz_f() {
+; CHECK-NEXT:    ret float 1.250000e+00
+;
+  %res = call float @llvm.nvvm.fmax.ftz.f(float 1.25, float -2.0)
```
Do we have any test cases that demonstrate that FTZ is actually in effect for those intrinsics (I.e. the cases where a non-FTZ variant of the intrinsic would return a different value for the same arguments?)

LewisCrawford replied:
Yes, I added the following cases for all the NVVM fmax/fmin intrinsics:

- FMax(1.25, -2.0) 
- FMax(+Subnormal, NaN) 
- FMax(NaN, undef)

- FMin(1.25, -2.0)
- FMin(+Subnormal, NaN)
- FMin(NaN, undef)

In the cases with Subnormal, the FTZ version returns 0.0, and the non-FTZ version returns the unmodified subnormal value 0x380FFFFFC0000000

LewisCrawford replied:
The inputs `(1.25, -2.0)` are only there to test that the intrinsics function as expected for regular (non-subnormal) floats, so the FTZ and non-FTZ variants are expected to behave identically for those inputs.

The FTZ behaviour should be covered already by the cases inputs `(+Subnormal, NaN)` and `(+Subnormal, undef)`.

I've added an additional set of inputs for `(+Subnormal, -Subnormal)` now to increase the coverage there even further, and showcase the behaviour of `-0.0 < +0.0` when the FTZ modifier makes the subnormals into zeroes.

Artem-B replied:
Thank you. 

> The inputs (1.25, -2.0) are only there to test that the intrinsics function as expected for regular (non-subnormal) floats, so the FTZ and non-FTZ variants are expected to behave identically for those inputs.

Agreed. It is useful, but it's the corner cases I would like to check. They tests you've already added cover most of the insteresting combinations, but we still seem to be missing the tests for `{normal FP, Subnormal}` combinations.

E.g. `fmax(0.0, +Subnormal)`.  FTZ variant would return 0.0, while non-FTZ would presumably pass through `+Subnormal`.

LewisCrawford replied:
I've added cases for `fmax(+Subnormal, 0.0)` and `fmin(-Subnormal, 0.0)` now too.

LewisCrawford committed:
Add missing nan modifier to ftz.nan tests

The tests for fmax/fmin.ftz.nan were missing the .nan modifier,
so were erroneously testing fmax/fmin.ftz twice.

This patch adds the missing modifier, and updates the expected
values to NaN where the instruction should propagate nan inputs.

LewisCrawford committed:
Add even more subnormal testing

LewisCrawford committed:
Add yet more subnormal tests

Add test cases for:
  fmax(+Subnormal, 0.0)
  fmin(-Subnormal, 0.0)

Artem-B reviewed (approved):
LGTM. Thank you!

LewisCrawford committed:
Merge branch 'main' into fold_nvvm_fmin_fmax
